frontend/test/pytest/test_callback.py

```diff
@@ -811,6 +812,330 @@ def wrapper_jit(x):
+
+    @some_func.bwd
+    def some_func_bws(res, dy):
+        x = res
+        return (jnp.array([jnp.cos(x), -jnp.sin(x)]) @ dy,)
+
+    @qml.qjit
+    @grad
+    def result(x):
+        return jnp.sum(some_func(jnp.sin(x)))
+
+    @jax.jit
+    @jax.grad
+    def expected(x):
+        x = jnp.sin(x)
+        return jnp.sin(x) + jnp.cos(x)
+
+    x = 0.435
+    assert np.allclose(result(x), expected(x))
+
+    # Array(0.4565774, dtype=float64)
+
+
+def test_scalar_in_tuple_scalar_array_out():
+    """Test scalar in tuple scalar array out"""
+
+    @pure_callback
+    def some_func(
+        x,
+    ) -> (jax.ShapeDtypeStruct(tuple(), jnp.float64), jax.ShapeDtypeStruct((2,), jnp.float64)):
+        y = x**2, np.array([np.sin(x), np.cos(x)])
+        return y
+
+    @some_func.fwd
+    def some_func_fwd(x):
+        return some_func(x), x
+
+    @some_func.bwd
+    def some_func_bwd(res, dy):
+        x = res
+        vjp0 = 2 * x * dy[0]
+        vjp1 = jnp.array([jnp.cos(x), -jnp.sin(x)]) @ dy[1]
+        return (vjp0 + vjp1,)
+
+    @qml.qjit
+    @grad
+    def result(x):
+        a, b = some_func(jnp.sin(x))
+        return a + jnp.sum(b)
+
+    @jax.jit
+    @jax.grad
+    def expected(x):
+        x = jnp.sin(x)
+        return jnp.sin(x) + jnp.cos(x) + x**2
+
+    x = 0.435
+    assert np.allclose(result(x), expected(x))
+
+    # Array(1.2209063, dtype=float64)
+
+
+def test_array_in_tuple_array_out():
+    """Test array in tuple array out"""
+
+    def _some_func(x):
+        return np.sin(x), x**2
+
+    @pure_callback
+    def some_func(
+        x,
+    ) -> (jax.ShapeDtypeStruct((2,), jnp.float64), jax.ShapeDtypeStruct((2,), jnp.float64)):
+        return np.sin(x), x**2
+
+    @some_func.fwd
+    def some_func_fwd(x):
+        return some_func(x), x
+
+    @some_func.bwd
+    def some_func_bws(res, dy):
+        x = res
+        vjp0 = jnp.cos(x) * dy[0]
+        vjp1 = 2 * x * dy[1]
+        return (vjp0 + vjp1,)
+
+    @qml.qjit
+    @grad
+    def result(x):
+        return jnp.dot(*some_func(jnp.sin(x)))
+
+    @jax.jit
+    @jax.grad
+    def expected(x):
+        x = jnp.sin(x)
+        return jnp.dot(jnp.sin(x), x**2)
+
+    x = jnp.array([1.0, 0.5])
+    assert np.allclose(result(x), expected(x))
+
+    # Array([0.9329284 , 0.56711537], dtype=float64)
+
+
+def test_tuple_array_in_tuple_array_out():
+    """Test tuple array in tuple array out"""
+
+    @pure_callback
+    def some_func(
+        x, y
+    ) -> (jax.ShapeDtypeStruct((2,), jnp.float64), jax.ShapeDtypeStruct((2,), jnp.float64)):
+        return np.sin(x) @ y, y**2
+
+    @some_func.fwd
+    def some_func_fwd(x, y):
+        return some_func(x, y), (x, y)
+
+    @some_func.bwd
+    def some_func_bwd(res, dy):
+        x, y = res
+        vjp0 = y * jnp.cos(x) * jnp.reshape(dy[0], (-1, 1))
+        vjp1 = dy[0] @ jnp.sin(x) + 2 * y * dy[1]
+        return (vjp0, vjp1)
+
+    @qml.qjit
+    @partial(grad, argnum=[0, 1])  # We just use argnum instead of argnums?
```

Comment:
oh, do we? Huh.

Reply:
Yes unfortunate convention we took from pennylane, we should definitely update this to be `argnums`.